allennlp/nn/beam_search.py

```diff
@@ -25,6 +27,365 @@
+        self.with_replacement = with_replacement
+
+    @overrides
+    def sample_nodes(
+        self, log_probs: torch.Tensor, per_node_beam_size: int, state: StateType
+    ) -> Tuple[torch.Tensor, torch.Tensor, StateType]:
+        if self.temperature != 1.0:
+            _probabilities = torch.nn.functional.softmax(log_probs / self.temperature, dim=-1)
+        else:
+            _probabilities = log_probs.exp()
+
+        selected_indices = torch.multinomial(
+            _probabilities, per_node_beam_size, replacement=self.with_replacement
+        )
+
+        return torch.gather(log_probs, 1, selected_indices), selected_indices, state
+
+
+@Sampler.register("top-k")
+class TopKSampler(Sampler):
+    """
+    Represents a `Sampler` which redistributes the probability mass function among
+    the top `k` choices then selects from that subset
+    `k` is the number of highest-probability options that the returned choice will be selected from.
+    `temperature` modules the probabilities of the selected tokens. A `temperature` below 1.0 produces a
+    sharper probability distribution and a `temperature` above 1.0 produces a flatter probability
+    distribution.
+    Registered as a `Sampler` with name "top-k".
+    """
+
+    def __init__(
+        self,
+        k: int = 1,
+        temperature: float = 1.0,
+        filter_val: float = -float("inf"),
+        with_replacement: bool = False,
+    ):
+        if not isinstance(k, int) or k < 1:
+            raise ValueError("k must be a positive integer")
+        self.k = k
+        self.temperature = temperature or 1.0
+        self.filter_val = min_value_of_dtype(torch.float)
+        self.with_replacement = with_replacement
+
+    @overrides
+    def sample_nodes(
+        self, log_probs: torch.Tensor, per_node_beam_size: int, state: StateType
+    ) -> Tuple[torch.Tensor, torch.Tensor, StateType]:
+        assert self.k <= len(log_probs[-1])
+
+        if self.temperature != 1.0:
+            _log_probs = torch.nn.functional.log_softmax(log_probs / self.temperature, dim=-1)
+        else:
+            _log_probs = log_probs
+
+        # Find the indices that are not to be selected from
+        min_threshold = torch.topk(_log_probs, self.k)[0][..., -1].unsqueeze(dim=-1)
+        filtered_indices = _log_probs < min_threshold
+
+        # Prevent the filtered indiceds from being selected
+        _log_probs[..., filtered_indices] = self.filter_val
+        filtered_probabilities = torch.nn.functional.softmax(_log_probs, dim=-1)
+
+        # Shape: (batch_size*beam_size, per_node_beam_size)
+        selected_indices = torch.multinomial(
+            filtered_probabilities, per_node_beam_size, replacement=self.with_replacement
+        )
+
+        # Return (selected log probabilities, selected classes)
+        # shape: (batch_size*beam_size, num_samples), (batch_size*beam_size, num_samples)
+        return torch.gather(_log_probs, 1, selected_indices), selected_indices, state
```

Comment:
I think we could avoid an extra call to softmax by doing something like this instead:

```suggestion
        # shape (both): (batch_size, k)
        top_k_log_probs, top_k_indices = log_probs.topk(self.k, dim=-1)

        # Apply temperature if necessary.
        # shape: (batch_size, k)
        if self.temperature != 1.0:
            top_k_log_probs = top_k_log_probs / self.temperature

        # Re-normalize the subset.
        # shape: (batch_size, k)
        normalized_top_k_probs = torch.nn.functional.softmax(top_k_log_probs, dim=-1)

        # Sample from the re-normalized subset.
        # NOTE: These indices are not indices into `log_probs`, they are indices into `top_k_log_probs`.
        # shape: (batch_size, per_node_beam_size)
        sampled_indices = torch.multinomial(
            normalized_top_k_probs, per_node_beam_size, replacement=self.replacement
        )

        # Convert `sampled_indices` back to indices in the original `log_probs` tensor.
        # shape: (batch_size, per_node_beam_size)
        indices = top_k_indices.gather(-1, sampled_indices)

        return log_probs.gather(1, indices), indices, state
```